### ISSUE TITLE
Future of libpd_wrapper

### ISSUE THREAD

**Issue by Spacechild1** [libpd/libpd#403]
Since @danomatika has already talked about this in various places, I think it might be a good idea to discuss this in a central place.

Since most (all?) sources for the libpd wrapper are already part of Pure Data itself, do we really need to keep libpd as a separate project at all? Once all sources are part of the Pure Data, the only thing that remains is the build system. However, Pure Data already contains two build systems for libpd:
1. a dedicated makefile in the `libpd` subfolder
2. as part of the autotools build system

I would suggest to fully integrate the libpd wrapper into Pure Data by moving the libpd makefile and CMakeLists.txt into `pure-data/libpd`, possibly replacing Pd's own libpd makefile.

The only thing that remains are the various language bindings. @danomatika already suggested a few times to split these into dedicated projects and ideally find maintainers for those he cannot maintain himself. I would totally agree with that!

**Comment by danomatika** [comment]
I decided to pull back for building the libpd sources in `src/pure-data` for now as it, quite frankly, slows down any current experimentation etc with the C layer when commits and discussion are lost within the mass of issues and PRs on the `pure-data` repo itself. For instance, claude's Emscripten PR would have to be held back / rejected as it relies on a small change to z_libpd.h. If we keep things are they are now, I can accept that PR, people using this repo can work with it directly, and we can fold the fixes made here after testing back to `pure-data` for the next release cycle.

**Comment by danomatika** [comment]
> However, Pure Data already contains two build systems for libpd:
> 
> 1. a dedicated makefile in the `libpd` subfolder
> 2. as part of the autotools build system

I would also note that:
* the Makefile here predated both of these,
* #1 -> I didn't write thisMakefile and it did not seem to be based on the libpd Makefile (although I have now been updating it)
* #2 -> I wasn't aware that this even existed before a couple days ago ;)

**Comment by danomatika** [comment]
Also, I mention this not as "sour grapes" but that I am not 100% all bases are covered with the two build systems mentioned when the libpd Makefile, Xcode project, etc which have been in used since circa 2010 / 2011.

**Comment by Spacechild1** [comment]
> as it, quite frankly, slows down any current experimentation etc with the C layer when commits and discussion are lost within the mass of issues and PRs on the pure-data repo itself. 

Fair enough. But wouldn't it make more sense to simply fork `pure-data`? There would be no duplication of source files or build systems and contributing back to upstream would be as simple as making a PR.

We can put all libpd related build systems to `pure-data/libpd`.

To be clear, I'm not saying we should do all of this in the near future, I'm more thinking about the long-term goal.

**Comment by danomatika** [comment]
Eh no, I would not want to add yet another fork. With the current situation, I can keep stable tags working with the `pure-data` submodule. I do *want* to transition away but am unsure now since it literally blocked the emscripten PR and another PR. I am trying to think what is the best workflow to blow through the masses of issues and PR at pure-data proper...

One solution would be `libpd` branches on `pure-data` such as:
* `libpd`: stable development, could be periodically merged into `develop` for the next pd release cycle
* `libpd/develop`: any unstable development cycles (probably not needed very often)
* `libpd/feature/blahblah`: feature or bugfixe branches, as needed

We just need a clear workflow which hopefully keeps things simple and easy. For instance, I currently do *not* subscribe the to `pure-data` repo issues as it's overload for me but I can partially deal with things here.

Again, I do want to move to stable C sources in `pure-data` I just had to second-guess myself about transitioning now. In any case, a transition would involve changing all included build systems to use `pure-data/src` while leaving `libpd_wrapper` there for 0.15 with notes about this. Then the dir can be removed or cleared in 0.16 or 0.17.

What do you think @umlaeute?

**Comment by Spacechild1** [comment]
> One solution would be libpd branches on pure-data such as:

I also thought about libpd branches features first. I would definitely prefer that. The problem is how to separate PRs/issues...

Of course, we can use tags to distinguish libpd related PRs and issues. However, I don't think GitHub has a mechanism to only subscribe to issues/PRs by tags. (GitLab apparently does have this feature!)

**Comment by danomatika** [comment]
OTOH C API changes at this point are realistically not going to happen very often and not without core developer involvement, so maybe I am overthinking things as a reaction to last nights feeling of "oh, Pd 0.55-2 is suddenly out, I just want to quickly release libpd 0.15 to track it... then trying to deal with all the issues and PRs again"

**Comment by danomatika** [comment]
See #404 for and updated the build system transition.

**Comment by umlaeute** [comment]
> * `libpd`: stable development, could be periodically merged into `develop` for the next pd release cycle
> * `libpd/develop`: any unstable development cycles (probably not needed very often)

this is of course not possible with git: but `libpd/main` would be a good branch name for stable development (just telling)

**Comment by danomatika** [comment]
> > * `libpd`: stable development, could be periodically merged into `develop` for the next pd release cycle
> > * `libpd/develop`: any unstable development cycles (probably not needed very often)
> 
> this is of course not possible with git: but `libpd/main` would be a good branch name for stable development (just telling)

Not possible? They are just branch names. It would be a place for people to submit PRs to for libpd without having to go into `pure-data` master.

**Comment by umlaeute** [comment]
> Not possible? They are just branch names

Well, `git` maps these branch names internally to a filesystem structure, with the actual branches being files.

Since you cannot have both a file named `libpd` and a file named `libpd/develop` (aka a file `develop` in a directory `libpd `), this making scheme is not possible. 
But as I said: I'm just telling. You can try (and prove I'm wrong)

**Comment by danomatika** [comment]
The more I think about it, this could work.

1. create a pure-data repo `libpd/main` branch
2. the libpd repo's `pure-data` submodule tracks this branch
3. update the submodule commit now and then for stable development
4. features, bug fixes, etc can go into `libpd/feature/..." and "libpd/bugfix/..." branches

I feel better if I can keep a running branch going as opposed to making a PR, then waiting to merge into pure-data `master` or `develop`. This would probably be flexible enough but would need to be communicated to people here.

**Comment by umlaeute** [comment]
> It does actually work, you just have to use `--`: `git checkout libpd --`

this just resolves the ambiguity between the [`./libpd`](https://github.com/pure-data/pure-data/tree/master/libpd) directory in the respository and a branch named `libpd`.
i was talking about the internal representation of branches: for each branch you have a file in `.git/refs/heads/`: if you branch is named `libpd`, you have a file `.git/refs/heads/libpd`; if your branch is named `libpd/main` you have a file  `.git/refs/heads/libpd/main`. since you cannot have both a *file*  `.git/refs/heads/libpd` and a file ` `.git/refs/heads/libpd/main` (because this requires a *directory* ` `.git/refs/heads/libpd`), git makes it impossible to have a repository with two branches `libpd` and `libpd/main`.

But anyhow:
> Ok, I changed the branch from `libpd` to `libpd/main`. 

which resolves the issue.


> Hopefully this doesn't affect anyone as it was only set for about 10 minutes.

we can probably be pretty sure that no human went into branch conflicts because of the renames, but *of course* a commit to the repository here automatically triggered a sync with other repositories, namely https://git.iem.at/pd/pure-data and https://gitlab.com/pure-data/pure-data (for the sake of the CI builds).
the mirroring typically cannot handle branch-renames like this, but afaict i have now fixed it manually.

**Comment by danomatika** [comment]
> > Ok I made a new one, but I hate to say that I don't remember which options were set. Do you?
> 
> no :-/ i was hoping you wrote them down before deleting the protection.

That would have been _smart_. I somehow assumed they were set to defaults which would appear on recreation. Doh.
